src/eos-updater-poll-common.c

```diff
@@ -312,6 +310,118 @@ get_booted_refspec (gchar               **booted_refspec,
   return TRUE;
 }
 
+static gboolean
+get_refspec_to_upgrade_on_from_deployment (OstreeSysroot     *sysroot,
+                                           OstreeDeployment  *booted_deployment,
+                                           gchar            **out_refspec_to_upgrade_from_deployment,
+                                           GError           **error)
+{
+  const gchar *checksum = ostree_deployment_get_csum (booted_deployment);
+  g_autoptr(GVariant) commit = NULL;
+  g_autoptr(GVariant) metadata = NULL;
+  g_autoptr(GVariant) ref_for_deployment_variant = NULL;
+  const gchar *refspec_for_deployment = NULL;
+  g_autoptr(OstreeRepo) repo = NULL;
+
+  g_return_val_if_fail (out_refspec_to_upgrade_from_deployment != NULL, FALSE);
+
+  if (!ostree_sysroot_get_repo (sysroot, &repo, NULL, error))
+   return FALSE;
+
+  if (!ostree_repo_load_variant (repo,
+                                 OSTREE_OBJECT_TYPE_COMMIT,
+                                 checksum,
+                                 &commit,
+                                 error))
+    return FALSE;
+
+  /* Look up the checkpoint target to see if there is one on this commit. */
+  metadata = g_variant_get_child_value (commit, 0);
+  ref_for_deployment_variant = g_variant_lookup_value (metadata,
+                                                       "eos.checkpoint-target",
+                                                       G_VARIANT_TYPE_STRING);
+
+  /* No metadata tag on this commit, just return TRUE with no value */
+  if (ref_for_deployment_variant == NULL)
+    {
+      *out_refspec_to_upgrade_from_deployment = NULL;
+      return TRUE;
+    }
+
+  refspec_for_deployment = g_variant_get_string (ref_for_deployment_variant, NULL);
+
+  if (refspec_for_deployment != NULL &&
```

Comment:
It’s not possible for `g_variant_get_string()` to return `NULL`, so the `refspec_for_deployment != NULL` condition can be dropped here.

Aside from that, this commit (`poll: Allow distributors to override the refspec to update from`). Looks good to me. If (and only if) it’s possible to do so without any rebase conflicts, it would be good to move this earlier in the branch, then I can cherry pick it and push it to master now. (Or you could split it out into a separate PR.) If it depends on all of the prior commits, don’t bother though.

Reply:
Also a good point. NULL-check removed.